File: src/schema.ts
import { z } from "zod"

export const TicketStatusSchema = z.enum(["open", "doing", "review", "blocked", "done"])
export type TicketStatus = z.infer<typeof TicketStatusSchema>

export const TicketPrioritySchema = z.string().min(1)
export type TicketPriority = z.infer<typeof TicketPrioritySchema>

export const NextActionTypeSchema = z.enum([
  "agent_action",
  "owner_gate",
  "release_gate",
  "blocked",
])

export const SourceSchema = z.record(z.string(), z.unknown()).nullable().default(null)

export const ArtifactSchema = z.union([
  z
    .object({
      type: z.string().default("artifact"),
      value: z.string().default(""),
      ts: z.string().default(""),
    })
    .passthrough(),
  z.string().transform((value) => ({ type: "artifact", value, ts: "" })),
])

export const LinkMapSchema = z
  .object({
    github_issues: z.array(z.string()).default([]),
    prs: z.array(z.string()).default([]),
    threads: z.array(z.string()).default([]),
    cron_jobs: z.array(z.string()).default([]),
  })
  .default({ github_issues: [], prs: [], threads: [], cron_jobs: [] })
export const LinkKindSchema = z.enum(["github_issues", "prs", "threads", "cron_jobs"])
export type LinkKind = z.infer<typeof LinkKindSchema>

export const LogEntrySchema = z
  .object({
    ts: z.string().default(""),
    action: z.string().optional(),
    note: z.string().optional(),
    phase: z.string().optional(),
    decision: z.string().optional(),
    evidence: z.string().optional(),
    blocker: z.string().optional(),
    next: z.string().optional(),
    next_action: z
      .object({
        type: NextActionTypeSchema.optional(),
        command: z.string().optional(),
        owner: z.string().optional(),
      })
      .optional(),
  })
  .passthrough()

export const CurrentSchema = z
  .object({
    phase: z.string().optional(),
    decision: z.string().optional(),
    evidence: z.string().optional(),
    blocker: z.string().optional(),
    next: z.string().optional(),
    note: z.string().optional(),
    next_action: z
      .object({
        type: NextActionTypeSchema.optional(),
        command: z.string().optional(),
        owner: z.string().optional(),
      })
      .optional(),
  })
  .passthrough()

export const TicketSchema = z
  .object({
    id: z.string().regex(/^T-\d{8}-\d{3}$/),
    title: z.string().min(1),
    status: TicketStatusSchema,
    priority: TicketPrioritySchema,
    type: z.string().min(1),
    source: SourceSchema,
    goal: z.string().default(""),
    acceptance: z.array(z.string()).default([]),
    tags: z.array(z.string()).default([]),
    artifacts: z.array(ArtifactSchema).default([]),
    links: LinkMapSchema,
    parent: z.string().nullable().default(null),
    children: z.array(z.string()).default([]),
    assignee: z.string().default("iyen"),
    created: z.string(),
    updated: z.string(),
    closed: z.string().nullable().default(null),
    log: z.array(LogEntrySchema).default([]),
    current: CurrentSchema.optional(),
  })
  .passthrough()
export type Ticket = z.infer<typeof TicketSchema>

export const IndexSchema = z
  .object({
    version: z.number().optional(),
    lastId: z.string().nullable().optional(),
    tickets: z.record(
      z.string(),
      z
        .object({
          title: z.string(),
          status: TicketStatusSchema,
          priority: TicketPrioritySchema,
          type: z.string(),
          updated: z.string(),
        })
        .passthrough(),
    ),
  })
  .passthrough()
export type TicketIndex = z.infer<typeof IndexSchema>

export const CreateTicketInputSchema = z.object({
  title: z.string().min(1),
  type: z.string().min(1).default("chore"),
  priority: TicketPrioritySchema.default("medium"),
  goal: z.string().default(""),
  parent: z.string().optional(),
  assignee: z.string().min(1).default("iyen"),
  acceptance: z.array(z.string()).default([]),
  tags: z.array(z.string()).default([]),
  source: z.string().optional(),
})
export type CreateTicketInput = z.infer<typeof CreateTicketInputSchema>

export const StatusInputSchema = z.object({
  id: z.string().regex(/^T-\d{8}-\d{3}$/),
  status: TicketStatusSchema.exclude(["done"]).or(z.literal("done")),
  artifact: z.string().optional(),
  evidence: z.string().optional(),
  note: z.string().optional(),
})
export type StatusInput = z.infer<typeof StatusInputSchema>

export const CheckpointInputSchema = z
  .object({
    id: z.string().regex(/^T-\d{8}-\d{3}$/),
    phase: z.string().optional(),
    decision: z.string().optional(),
    evidence: z.string().optional(),
    blocker: z.string().optional(),
    next: z.string().optional(),
    note: z.string().optional(),
    nextType: NextActionTypeSchema.optional(),
    nextCommand: z.string().optional(),
    nextOwner: z.string().optional(),
  })
  .refine(
    (input) =>
      input.phase !== undefined ||
      input.decision !== undefined ||
      input.evidence !== undefined ||
      input.blocker !== undefined ||
      input.next !== undefined ||
      input.note !== undefined ||
      input.nextType !== undefined ||
      input.nextCommand !== undefined ||
      input.nextOwner !== undefined,
    { message: "checkpoint requires at least one field" },
  )
export type CheckpointInput = z.infer<typeof CheckpointInputSchema>

export const LinkInputSchema = z.object({
  id: z.string().regex(/^T-\d{8}-\d{3}$/),
  kind: LinkKindSchema,
  value: z.string().min(1),
})
export type LinkInput = z.infer<typeof LinkInputSchema>

export const AddLogInputSchema = z.object({
  id: z.string().regex(/^T-\d{8}-\d{3}$/),
  note: z.string().min(1),
})
export type AddLogInput = z.infer<typeof AddLogInputSchema>

File: rust-ticket-flow/crates/ticket-flow-cli/src/mcp/schema.rs
use serde_json::{Value, json};

pub(crate) fn tool_definitions() -> Value {
    json!([
        tool(
            "ticket_create",
            "Create an agent ticket.",
            object(
                json!({
                    "title": string("Ticket title"),
                    "type": string("Ticket type"),
                    "priority": string("Ticket priority"),
                    "goal": string("Goal statement"),
                    "assignee": string("Ticket owner"),
                    "acceptance": string_array("Acceptance criteria"),
                    "tags": string_array("Tags"),
                    "parent": string("Parent ticket ID"),
                    "source": string("Source reference as kind:ref")
                }),
                &["title"],
            )
        ),
        tool(
            "ticket_list",
            "List active agent tickets.",
            object(json!({"status": status_enum()}), &[],)
        ),
        tool(
            "ticket_get",
            "Read an active or archived ticket.",
            id_schema()
        ),
        tool(
            "ticket_import",
            "Import a source ticket store into ticket-flow.",
            object(
                json!({"sourceRoot": string("Source ticket store root")}),
                &["sourceRoot"],
            )
        ),
        tool(
            "ticket_update_status",
            "Apply a ticket status transition.",
            object(
                json!({
                    "id": string("Ticket ID"),
                    "status": status_enum(),
                    "artifact": string("Artifact value for review/done transitions"),
                    "evidence": string("Evidence artifact value"),
                    "note": string("Status transition note")
                }),
                &["id", "status"],
            )
        ),
        tool(
            "ticket_link",
            "Attach an external reference to a ticket.",
            object(
                json!({
                    "id": string("Ticket ID"),
                    "kind": {"type":"string","enum":["github_issues","prs","threads","cron_jobs"]},
                    "value": string("Reference value")
                }),
                &["id", "kind", "value"],
            )
        ),
        tool(
            "ticket_add_log",
            "Append a note log entry to a ticket.",
            object(
                json!({"id": string("Ticket ID"), "note": string("Log note")}),
                &["id", "note"],
            )
        ),
        tool(
            "ticket_checkpoint",
            "Write the current checkpoint payload.",
            object(
                json!({
                    "id": string("Ticket ID"),
                    "phase": string("Current phase"),
                    "decision": string("Current decision"),
                    "evidence": string("Evidence summary"),
                    "blocker": string("Blocker summary"),
                    "next": string("Next step summary"),
                    "note": string("Checkpoint note"),
                    "nextType": {"type":"string","enum":["agent_action","owner_gate","release_gate","blocked"]},
                    "nextCommand": string("Agent command"),
                    "nextOwner": string("Next owner")
                }),
                &["id"],
            )
        ),
        tool(
            "ticket_agent_actions",
            "List tickets with agent_action next actions.",
            object(json!({}), &[],)
        )
    ])
}

fn tool(name: &str, description: &str, input_schema: Value) -> Value {
    json!({ "name": name, "description": description, "inputSchema": input_schema })
}

fn id_schema() -> Value {
    object(json!({"id": string("Ticket ID")}), &["id"])
}

fn object(properties: Value, required: &[&str]) -> Value {
    json!({
        "type": "object",
        "properties": properties,
        "required": required,
        "additionalProperties": false
    })
}

fn string(description: &str) -> Value {
    json!({ "type": "string", "description": description })
}

fn string_array(description: &str) -> Value {
    json!({ "type": "array", "items": {"type":"string"}, "description": description })
}

fn status_enum() -> Value {
    json!({ "type":"string", "enum":["open","doing","review","blocked","done"] })
}
